dimod/include/dimod/constrained_quadratic_model.h

```diff
@@ -0,0 +1,493 @@
+        // parent_ pointers are pointing at the correct object.
+
+        swap(this->objective, other.objective);
+        this->objective.parent_ = this;
+        other.objective.parent_ = &other;
+
+        swap(this->constraints_, other.constraints_);
+        for (auto& c : this->constraints_) {
+            c.parent_ = this;
+        }
+        for (auto& c : other.constraints_) {
+            c.parent_ = &other;
+        }
+
+        swap(this->varinfo_, other.varinfo_);
+    }
+
+    std::vector<Constraint<bias_type, index_type>> constraints_;
+
+    struct varinfo_type {
+        Vartype vartype;
+        bias_type lb;
+        bias_type ub;
+
+        varinfo_type(Vartype vartype, bias_type lb, bias_type ub)
+                : vartype(vartype), lb(lb), ub(ub) {
+            assert(lb <= ub);
+
+            assert(lb >= vartype_info<bias_type>::min(vartype));
+            assert(ub <= vartype_info<bias_type>::max(vartype));
+
+            assert(vartype != Vartype::BINARY || lb == 0);
+            assert(vartype != Vartype::BINARY || ub == 1);
+
+            assert(vartype != Vartype::SPIN || lb == -1);
+            assert(vartype != Vartype::SPIN || ub == +1);
+        }
+
+        explicit varinfo_type(Vartype vartype)
+                : vartype(vartype),
+                  lb(vartype_info<bias_type>::default_min(vartype)),
+                  ub(vartype_info<bias_type>::default_max(vartype)) {}
+    };
+
+    std::vector<varinfo_type> varinfo_;
+};
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>::ConstrainedQuadraticModel()
+        : objective(this), constraints_(), varinfo_() {}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>::ConstrainedQuadraticModel(
+        const ConstrainedQuadraticModel& other)
+        : objective(other.objective), constraints_(other.constraints_), varinfo_(other.varinfo_) {
+    objective.parent_ = this;
+    for (auto& c : constraints_) {
+        c.parent_ = this;
+    }
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>::ConstrainedQuadraticModel(
+        ConstrainedQuadraticModel&& other) noexcept
+        : ConstrainedQuadraticModel() {
+    swap(*this, other);
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>&
+ConstrainedQuadraticModel<bias_type, index_type>::operator=(ConstrainedQuadraticModel other) {
+    swap(*this, other);
+    return *this;
+}
+
+template <class bias_type, class index_type>
+ConstrainedQuadraticModel<bias_type, index_type>&
+ConstrainedQuadraticModel<bias_type, index_type>::operator=(
+        ConstrainedQuadraticModel&& other) noexcept {
+    swap(*this, other);
+    return *this;
+}
+
+template <class bias_type, class index_type>
+index_type ConstrainedQuadraticModel<bias_type, index_type>::add_constraint() {
```

Comment:
We could do away with this function, just lets the option of adding 100s of empty constraints. I do not see any value of having this function.

Reply:
I use it :smile: